src/Data/Id.hs

```diff
@@ -4,69 +4,172 @@
 
-instance PathPiece UUID.UUID where
-  fromPathPiece = UUID.fromText
-  toPathPiece = UUID.toText
+#ifndef ghcjs_HOST_OS
 
-makePrisms ''Id
+instance Csv.FromField (Id tag) where
+  parseField = parseField
+    >=> fmap mkId . maybe (fail "invalid UUID in CSV") pure . UUID.fromText
+
+instance Csv.ToField (Id tag) where
+  toField = Csv.toField . UUID.toText . coerce
+
+#endif
 
 -- | This is a more \"explicit\" 'coerce' specifically for 'Id'.
 -- You are forced to explicitly specify the phantom types you are converting
 -- via the @TypeApplications@ compiler extension.
-coerceId :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
+coerceId, unsafeIdTagConvert :: forall a b. Id (Ambiguous a) -> Id (Ambiguous b)
 coerceId = coerce
 {-# INLINE coerceId #-}
 
+-- just another name for 'coerceId'
+unsafeIdTagConvert = coerce
+{-# INLINE unsafeIdTagConvert #-}
+
 -- | Id corresponding to UUID nil. Useful for testing etc.
 nilId :: forall a.  Id (Ambiguous a)
 nilId = Id UUID.nil
 {-# INLINE nilId #-}
 -- https://kcsongor.github.io/ambiguous-tags/
 type family Ambiguous (a :: k) :: j where
   Ambiguous x = x
+
+----------------- IntId
+newtype IntId t = IntId { unIntId :: Integer }
+  deriving stock Data
+  deriving newtype
+  ( Eq, Ord, Binary
+#ifndef ghcjs_HOST_OS
+  , PG.ToField, PG.FromField, Csv.ToField, Csv.FromField
+#endif
+  , FromJSON, ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema
+  , ToParamSchema, FromHttpApiData, ToHttpApiData, PathPiece, Flat, Arbitrary
+  , Num, Integral, Real, Enum )
+
+type role IntId nominal
+
+makePrisms ''IntId
+
+mkIntId :: forall s. Integer -> IntId s
+mkIntId = coerce
+{-# INLINE mkIntId #-}
+
+instance KnownSymbol s => Show (IntId s) where
+  show (IntId v) = "IntId-" <> symbolVal (Proxy @s) <> "-" <> show v
+
+-- | This is a more \"explicit\" 'coerce' specifically for 'IntId'.
+-- You are forced to explicitly specify the phantom types you are converting
+-- via the @TypeApplications@ compiler extension.
+coerceIntId, unsafeIntIdTagConvert
+  :: forall a b. IntId (Ambiguous a) -> IntId (Ambiguous b)
+coerceIntId = coerce
+{-# INLINE coerceIntId #-}
+
+-- just another name for 'coerceIntId'
+unsafeIntIdTagConvert = coerce
+{-# INLINE unsafeIntIdTagConvert #-}
+
+----------------- Name
+newtype Name t = Name { unName :: Text }
+  deriving stock Data
+  deriving newtype
+  ( Eq, Ord, Binary
+#ifndef ghcjs_HOST_OS
+  , PG.ToField, PG.FromField, Csv.ToField, Csv.FromField
+#endif
+  , FromJSON, ToJSON, NFData, Hashable, FromJSONKey, ToJSONKey, ToSchema
+  , ToParamSchema, FromHttpApiData, ToHttpApiData, PathPiece, Flat
+  , IsString, Semigroup, Monoid )
+
+type role Name nominal
+
+makePrisms ''Name
+
+mkName :: forall s. Text -> Name s
+mkName = coerce
+{-# INLINE mkName #-}
+
+instance KnownSymbol s => Show (Name s) where
+  show (Name v) = "Name-" <> symbolVal (Proxy @s) <> "-" <> show v
+
+instance Arbitrary (Name s) where
+  arbitrary = coerce . T.pack <$> arbitrary
```

Comment:
I think this would produce invalid characters. Might we want to use something like [`PrintableString`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#t:PrintableString)?